File: eco-core/core-plugin/src/main/java/com/willfp/ecoenchants/EcoEnchantsPlugin.java
package com.willfp.ecoenchants;

import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.eco.core.fast.FastItemStack;
import com.willfp.eco.core.integrations.IntegrationLoader;
import com.willfp.eco.util.TelekinesisUtils;
import com.willfp.ecoenchants.command.CommandEcoEnchants;
import com.willfp.ecoenchants.command.CommandEnchantinfo;
import com.willfp.ecoenchants.config.CustomEnchantsYml;
import com.willfp.ecoenchants.config.RarityYml;
import com.willfp.ecoenchants.config.TargetYml;
import com.willfp.ecoenchants.config.VanillaEnchantsYml;
import com.willfp.ecoenchants.display.EnchantDisplay;
import com.willfp.ecoenchants.enchantments.EcoEnchant;
import com.willfp.ecoenchants.enchantments.EcoEnchants;
import com.willfp.ecoenchants.enchantments.custom.CustomEcoEnchantRequirementListeners;
import com.willfp.ecoenchants.enchantments.custom.CustomEnchantEnableListeners;
import com.willfp.ecoenchants.enchantments.custom.CustomEnchantLookup;
import com.willfp.ecoenchants.enchantments.support.merging.anvil.AnvilListeners;
import com.willfp.ecoenchants.enchantments.support.merging.grindstone.GrindstoneListeners;
import com.willfp.ecoenchants.enchantments.support.obtaining.EnchantingListeners;
import com.willfp.ecoenchants.enchantments.support.obtaining.LootPopulator;
import com.willfp.ecoenchants.enchantments.support.obtaining.VillagerListeners;
import com.willfp.ecoenchants.enchantments.util.ItemConversions;
import com.willfp.ecoenchants.enchantments.util.TimedRunnable;
import com.willfp.ecoenchants.enchantments.util.WatcherTriggers;
import com.willfp.ecoenchants.integrations.mythicmobs.MythicMobsManager;
import com.willfp.ecoenchants.integrations.mythicmobs.plugins.IntegrationMythicMobs;
import com.willfp.ecoenchants.integrations.registration.RegistrationManager;
import com.willfp.ecoenchants.integrations.registration.plugins.IntegrationCMI;
import com.willfp.ecoenchants.integrations.registration.plugins.IntegrationEssentials;
import com.willfp.libreforge.LibReforgePlugin;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

@SuppressWarnings("unused")
public class EcoEnchantsPlugin extends LibReforgePlugin {
    /**
     * Instance of the plugin.
     */
    private static EcoEnchantsPlugin instance;

    /**
     * Rarity.yml.
     */
    private final RarityYml rarityYml;

    /**
     * Target.yml.
     */
    private final TargetYml targetYml;

    /**
     * VanillaEnchants.yml.
     */
    private final VanillaEnchantsYml vanillaEnchantsYml;

    /**
     * CustomEnchants.yml.
     */
    private final CustomEnchantsYml customEnchantsYml;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public EcoEnchantsPlugin() {
        instance = this;

        rarityYml = new RarityYml(this);
        targetYml = new TargetYml(this);
        vanillaEnchantsYml = new VanillaEnchantsYml(this);
        customEnchantsYml = new CustomEnchantsYml(this);

        this.registerJavaHolderProvider(player -> new ArrayList<>(CustomEnchantLookup.provideLevels(player)));
    }

    @Override
    public void handleEnableAdditional() {
        this.getLogger().info(EcoEnchants.values().size() + " Enchantments Loaded");

        TelekinesisUtils.registerTest(player -> FastItemStack.wrap(player.getInventory().getItemInMainHand()).getEnchantmentLevel(EcoEnchants.TELEKINESIS, false) > 0);
    }

    @Override
    public void handleDisableAdditional() {
        for (World world : Bukkit.getServer().getWorlds()) {
            world.getPopulators().removeIf(blockPopulator -> blockPopulator instanceof LootPopulator);
        }
    }

    @Override
    public void handleReloadAdditional() {
        this.getDisplayModule().update();
        for (EcoEnchant enchant : EcoEnchants.values()) {
            HandlerList.unregisterAll(enchant);
            this.getScheduler().runLater(() -> {
                if (enchant.isEnabled()) {
                    this.getEventManager().registerListener(enchant);

                    if (enchant instanceof TimedRunnable) {
                        this.getScheduler().syncRepeating((TimedRunnable) enchant, 5, ((TimedRunnable) enchant).getTime());
                    }
                }
            }, 1);
        }
        this.getScheduler().runTimer(() -> {
            for (EcoEnchant enchant : EcoEnchants.values()) {
                enchant.clearCachedRequirements();
            }
        }, 300, 300);
    }

    @Override
    protected void handleAfterLoad() {
        if (this.getConfigYml().getBool("loot.enabled")) {
            for (World world : Bukkit.getServer().getWorlds()) {
                world.getPopulators().removeIf(blockPopulator -> blockPopulator instanceof LootPopulator);
                world.getPopulators().add(new LootPopulator(this));
            }
        }
    }

    @Override
    @NotNull
    public List<IntegrationLoader> loadAdditionalIntegrations() {
        return Arrays.asList(
                new IntegrationLoader("Essentials", () -> RegistrationManager.register(new IntegrationEssentials())),
                new IntegrationLoader("CMI", () -> RegistrationManager.register(new IntegrationCMI())),
                new IntegrationLoader("MythicMobs", () -> MythicMobsManager.register(new IntegrationMythicMobs()))
        );
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandEnchantinfo(this),
                new CommandEcoEnchants(this)
        );
    }

    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new EnchantingListeners(this),
                new GrindstoneListeners(this),
                new AnvilListeners(this),
                new WatcherTriggers(this),
                new VillagerListeners(this),
                new ItemConversions(this),
                new CustomEnchantEnableListeners(this),
                new CustomEcoEnchantRequirementListeners(this)
        );
    }

    @Override
    @Nullable
    protected DisplayModule createDisplayModule() {
        return new EnchantDisplay(this);
    }

    @Override
    @NotNull
    public EnchantDisplay getDisplayModule() {
        return (EnchantDisplay) Objects.requireNonNull(super.getDisplayModule());
    }

    @Override
    public String getMinimumEcoVersion() {
        return "6.35.1";
    }

    /**
     * Get the instance of EcoEnchants.
     * <p>
     * Bad practice to use this.
     *
     * @return The instance.
     */
    public static EcoEnchantsPlugin getInstance() {
        return instance;
    }

    /**
     * Get rarity.yml.
     *
     * @return rarity.yml.
     */
    public RarityYml getRarityYml() {
        return this.rarityYml;
    }

    /**
     * Get target.yml.
     *
     * @return target.yml.
     */
    public TargetYml getTargetYml() {
        return this.targetYml;
    }

    /**
     * Get vanillaenchants.yml.
     *
     * @return vanillaenchants.yml.
     */
    public VanillaEnchantsYml getVanillaEnchantsYml() {
        return this.vanillaEnchantsYml;
    }

    /**
     * Get customenchants.yml
     *
     * @return customenchants.yml.
     */
    public CustomEnchantsYml getCustomEnchantsYml() {
        return customEnchantsYml;
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoenchants/EcoEnchantsPlugin.kt
package com.willfp.ecoenchants

import com.willfp.eco.core.anvil.AnvilHandlers
import com.willfp.eco.core.anvil.AnvilSettings
import com.willfp.eco.core.bstats.EcoMetricsChart
import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.display.DisplayModule
import com.willfp.eco.core.integrations.IntegrationLoader
import com.willfp.ecoenchants.commands.CommandEcoEnchants
import com.willfp.ecoenchants.commands.CommandEnchant
import com.willfp.ecoenchants.commands.CommandEnchantInfo
import com.willfp.ecoenchants.config.RarityYml
import com.willfp.ecoenchants.config.TargetsYml
import com.willfp.ecoenchants.config.TypesYml
import com.willfp.ecoenchants.config.VanillaEnchantsYml
import com.willfp.ecoenchants.display.DisplayCache
import com.willfp.ecoenchants.display.EnchantDisplay
import com.willfp.ecoenchants.display.EnchantSorter
import com.willfp.ecoenchants.enchant.EcoEnchantLevel
import com.willfp.ecoenchants.enchant.EcoEnchants
import com.willfp.ecoenchants.enchant.EnchantGUI
import com.willfp.ecoenchants.enchant.LoreConversion
import com.willfp.ecoenchants.enchant.registration.EnchantmentRegisterer
import com.willfp.ecoenchants.enchant.registration.ModernEnchantmentRegistererProxy
import com.willfp.ecoenchants.integrations.EnchantRegistrations
import com.willfp.ecoenchants.integrations.plugins.CMIIntegration
import com.willfp.ecoenchants.integrations.plugins.EssentialsIntegration
import com.willfp.ecoenchants.libreforge.EffectApplyRandomEnchant
import com.willfp.ecoenchants.mechanics.EcoEnchantsAnvilHandler
import com.willfp.ecoenchants.mechanics.EnchantingTableSupport
import com.willfp.ecoenchants.mechanics.ExtraItemSupport
import com.willfp.ecoenchants.mechanics.GrindstoneSupport
import com.willfp.ecoenchants.mechanics.LootSupport
import com.willfp.ecoenchants.mechanics.VillagerSupport
import com.willfp.ecoenchants.rarity.EnchantmentRarities
import com.willfp.ecoenchants.target.EnchantFinder
import com.willfp.ecoenchants.target.EnchantFinder.clearEnchantmentCache
import com.willfp.ecoenchants.target.EnchantmentTargets
import com.willfp.ecoenchants.type.EnchantmentTypes
import com.willfp.libreforge.NamedValue
import com.willfp.libreforge.effects.Effects
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.registerHolderPlaceholderProvider
import com.willfp.libreforge.registerHolderProvider
import com.willfp.libreforge.registerSpecificRefreshFunction
import org.bukkit.entity.LivingEntity
import org.bukkit.event.Listener

internal lateinit var plugin: EcoEnchantsPlugin
    private set

class EcoEnchantsPlugin : LibreforgePlugin() {
    val targetsYml = TargetsYml(this)
    val rarityYml = RarityYml(this)
    val typesYml = TypesYml(this)
    val vanillaEnchantsYml = VanillaEnchantsYml(this)
    var isLoaded = false
        private set

    val enchantmentRegisterer: EnchantmentRegisterer = this.getProxy(ModernEnchantmentRegistererProxy::class.java)

    init {
        plugin = this

        plugin.getProxy(ModernEnchantmentRegistererProxy::class.java).replaceRegistry()
    }

    override fun loadConfigCategories(): List<ConfigCategory> {
        return listOf(
            EcoEnchants
        )
    }

    override fun handleEnable() {
        Effects.register(EffectApplyRandomEnchant)

        registerHolderProvider(EnchantFinder.toHolderProvider())

        registerSpecificRefreshFunction<LivingEntity> {
            it.clearEnchantmentCache()
        }

        registerHolderPlaceholderProvider<EcoEnchantLevel> { it, _ ->
            listOf(
                NamedValue("level", it.level),
            )
        }

        registerAnvilHandler()
    }

    override fun handleAfterLoad() {
        isLoaded = true

        plugin.getProxy(ModernEnchantmentRegistererProxy::class.java).replaceRegistry()
    }

    override fun handleReload() {
        DisplayCache.reload()
        EnchantSorter.reload()
        ExtraItemSupport.reload()
        EnchantGUI.reload()

        registerAnvilHandler()
    }

    private fun registerAnvilHandler() {
        AnvilHandlers.register(
            EcoEnchantsAnvilHandler(),
            AnvilSettings(
                costExponent = configYml.getDouble("anvil.cost-exponent"),
                enchantLimit = configYml.getInt("anvil.enchant-limit"),
                useReworkPenalty = configYml.getBool("anvil.use-rework-penalty"),
                maxRepairCost = configYml.getInt("anvil.max-repair-cost"),
                clampRepairCost = configYml.getBool("anvil.clamp-repair-cost"),
                colorNameAllowed = { it.hasPermission("ecoenchants.anvil.color") }
            )
        )
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            VillagerSupport,
            EnchantingTableSupport,
            LootSupport,
            LoreConversion,
            GrindstoneSupport
        )
    }

    override fun loadIntegrationLoaders(): List<IntegrationLoader> {
        return listOf(
            IntegrationLoader("Essentials") { EnchantRegistrations.register(EssentialsIntegration) },
            IntegrationLoader("CMI") { EnchantRegistrations.register(CMIIntegration) }
        )
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandEcoEnchants,
            CommandEnchantInfo,
            CommandEnchant
        )
    }

    override fun loadDisplayModules(): List<DisplayModule> {
        if (!this.configYml.getBool("display.enabled")) {
            return emptyList()
        }

        return listOf(
            EnchantDisplay
        )
    }

    override fun getCustomCharts() = listOf(
        EcoMetricsChart.SingleLine("total_enchants") { EcoEnchants.values().size },
        EcoMetricsChart.SingleLine("total_rarities") { EnchantmentRarities.values().size },
        EcoMetricsChart.SingleLine("total_targets") { EnchantmentTargets.values().size },
        EcoMetricsChart.SingleLine("total_types") { EnchantmentTypes.values().size },
        EcoMetricsChart.SimplePie("display_enabled") {
            if (configYml.getBool("display.enabled")) "enabled" else "disabled"
        }
    )
}
